cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetTableMigrator.java

```diff
@@ -0,0 +1,175 @@
+
+                // the hbase version for all the columns are equal since every column of a row corresponds to one
+                // partition and a partition is created entirely within a single transaction.
+                hBaseVersion = timeStamp;
+              }
+            }
+          }
+
+          Preconditions.checkNotNull(hBaseVersion,
+                                     "There should have been at least one column in the scan. Table: {}, Rowkey: {}",
+                                     oldPartitionsTableId, result.getRow());
+
+          // additionally since 3.1.0, we keep two addition columns for each partition: creation time of the partition
+          // and the transaction write pointer of the transaction in which the partition was added
+          byte[] hbaseVersionBytes = Bytes.toBytes(hBaseVersion);
+          put.add(columnFamily, PartitionedFileSetDataset.WRITE_PTR_COL,
+                  hBaseVersion, hbaseVersionBytes);
+          // here, we make the assumption that dropping the six right-most digits of the transaction write pointer
+          // yields the timestamp at which it started
+          byte[] creationTimeBytes = Bytes.toBytes(hBaseVersion / MILLION);
+          put.add(columnFamily, PartitionedFileSetDataset.CREATION_TIME_COL,
+                  hBaseVersion, creationTimeBytes);
+
+          newDataTable.put(put);
+
+          // index the data table on the two columns
+          Put indexTableWritePointer = new Put(IndexedTable.createIndexKey(result.getRow(),
+                                                                           PartitionedFileSetDataset.WRITE_PTR_COL,
+                                                                           hbaseVersionBytes));
+          indexTableWritePointer.add(columnFamily, IndexedTable.IDX_COL, hBaseVersion, result.getRow());
+
+          Put indexTableCreationTime =
+            new Put(IndexedTable.createIndexKey(result.getRow(), PartitionedFileSetDataset.CREATION_TIME_COL,
+                                                creationTimeBytes));
+          indexTableCreationTime.add(columnFamily, IndexedTable.IDX_COL, hBaseVersion, result.getRow());
+
+          newIndexTable.put(ImmutableList.of(indexTableCreationTime, indexTableWritePointer));
+
+          LOG.debug("Deleting old key {}.", Bytes.toString(result.getRow()));
+          oldTable.delete(new Delete(result.getRow()));
+        }
+      } finally {
+        oldTable.close();
+        newIndexTable.close();
+        newDataTable.close();
+      }
+      LOG.info("Successfully migrated data from table {} Now deleting it.", Bytes.toString(oldTable.getTableName()));
+      hBaseAdmin.deleteTable(oldTable.getTableName());
+      LOG.info("Succsefully deleted old data table {}", Bytes.toString(oldTable.getTableName()));
+    } catch (Exception e) {
+      LOG.error("Error while migrating data from table: {}", oldPartitionsTableId, e);
```

Comment:
Actually, this can happen. In the very rare case that a dataset embeds two PFS, the PFSUpgrader first rewrites the spec and stores it in MDS, then attempts to migrate both embedded datasets. If the second one fails, then the user retries the upgrade, it will not even attempt to migrate the embedded tables again.